File: 0046-permutations/0046-permutations.cpp
class Solution {
public:
    void solve(int pos, int n, vector<int>& nums, vector<vector<int>>& ans){
        if (pos == n) {
            ans.push_back(nums);
            return;
        }
        for (int i = pos; i < n; i++) {
            swap(nums[i], nums[pos]);
            solve(pos + 1, n, nums, ans);
            swap(nums[i], nums[pos]);
        }
    }
    vector<vector<int>> permute(vector<int>& nums) {
        int len = nums.size();
        sort(nums.begin(), nums.end());
        vector<vector<int>> ans;
        solve(0, len, nums, ans);
        return ans;
    }
};

File: 0046-permutations/0046-permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []
        
        def dfs(pos):
            if pos == len(nums):
                # res.append(nums.copy())
                res.append(nums[:])
                return
            for i in range(pos, len(nums)):
                nums[i], nums[pos] = nums[pos], nums[i]
                dfs(pos + 1)
                nums[i], nums[pos] = nums[pos], nums[i]
                
        dfs(0)
        return res
        
# Time: O(N*N!)
# Space: O(N!)